### DISCUSSION TITLE
Add Offset Bleed Margins to existing PDF

### DISCUSSION THREAD

**Discussion by byteinsight** [pymupdf/PyMuPDF#2838] [Q&A] [answered]
Hi,

I have some pdfs that I need to add bleed margins to.   I have seen [this previous question](https://github.com/pymupdf/PyMuPDF/issues/436) which has got me some way towards a solution but not quite all the way there.   

I need to add a variable called bleed_margin_pixels to the top, bottom and outer edge.  At the moment I have

```
def resize_pages(self, bleed_margin_pixels):
        # Open the same file but with as a different name as our bleed document
        bleed_doc = fitz.open(self.indoc.name)

       # For each page
        for page_index in range(0, len(bleed_doc)):

            # Get the page rectangle and resize it to create rect_new
            page_rect = bleed_doc[page_index].rect
            new_page_rect = page_rect + (0, 0, bleed_margin_pixels, bleed_margin_pixels*2)

            # Offset Content holder for original page: 0 is for even pages else odd pages
            if page_index % 2 == 0:
                rect = fitz.Rect(0, bleed_margin_pixels, new_page_rect.width, new_page_rect.height-bleed_margin_pixels)
            else:
                rect = fitz.Rect(-bleed_margin_pixels, bleed_margin_pixels, new_page_rect.width, new_page_rect.height-bleed_margin_pixels)

            # Put the larger page behind the current one
            new_page = bleed_doc.new_page(page_index + 1, width=new_page_rect.width, height=new_page_rect.height)
            new_page.show_pdf_page(new_page_rect, self.indoc, page_index, clip=rect)

            # Delete old small page
            bleed_doc.delete_page(page_index)  


        return bleed_doc
```
        
But its the two lines:

```
rect = fitz.Rect(0, bleed_margin_pixels, new_page_rect.width, new_page_rect.height-bleed_margin_pixels)
rect = fitz.Rect(-bleed_margin_pixels, bleed_margin_pixels, new_page_rect.width, new_page_rect.height-bleed_margin_pixels)
```

where I'm kinda guessing and not sure its yet right.   I don't want the original content to be resized.  It should stay the same size.  

Thanks for your help.  :)

**Comment by JorjMcKie** [comment]
This is no issue, but a **_Discussions_** item. Transferring ...

**Comment by JorjMcKie** [comment]
Several comments:
1. If all you want is to reduce the **_visible part_** of a page, you can do this by simply setting one of the pages optional "XXXBox" rectangles, for instance the `CropBox`. This requires no copying to other pages or whatever. Just use `page.set_cropbox(newrect)`. Same is possible for `ArtBox`, `BleedBox`.
2. I am not sure if I am making a valid point, but deleting a some page **_renumbers_** all pages thereafter. If you delete page number 4, then the old number 5 will become number 4, 6 will now be 5 etc. I haven't following your code enough, but maybe you weren't aware of that. Anyway, this should not be necessary at all anyway.

**Comment by byteinsight** [comment]
Hi Jorj,     Thank you for your response and moving my question into the correct area.  I really appreciate your time. 

For clairfication, I don't need to reduce the visible part of the page but add 'whitespace' around it.  Although I appreciate this might be what you are describing from a different direction!   I've attached an image to show what I'd like to achieve.   

- The Blue Box is the original page where all the Rect, MBox and Crp Box box sizes are all 360 x 576.  
- The Red Box shows how I would like to increase the margins and overall page size to 369 x 594.  
- The 9 pixel growth on the width should occur on the outer edge. 
- The original content should remain at the orignal dimensions.    

The green box some what complicates this as its artwork I'm positioning as shown.   At the moment I'm able to add the green artwork to the PDF but on the original blue dimensions.   Trimming this once printed offsets the blue content hence the requirement for the red bleed area.

Hope this makes sense.   Thanks again for your time. 

![BleedDiagram](https://github.com/pymupdf/PyMuPDF/assets/5509451/66144a8b-8356-4882-b051-e8ce66f3a231)

**Reply by JorjMcKie** [reply, marked as answer]
Aha! Thanks, that is much clearer now! So we have the following rectangles:
1. Current page size `cur_rect = fitz.Rect(0, 0, 360, 576)`
2. New page size `new_rect = fitz.Rect(0, 0, 369, 594)`
3. Target rect on even (left) pages `evn_rect = fitz.Rect(9, 9, 369, 585)`
4. Target rect on odd (right) pages `odd_rect = fitz.Rect(0, 9, 360, 585)`

I hope I didn't make stupid errors. Rectangles 3 and 4 correspond the blue rectangles in your nice drawings.
So your script would open a new PDF and read the old, current PDF page by page, make a new page on the new PDF with size `new_rect` and position the current old page on either `evn_rect` or `odd_rect` within it, depending on the page number.
When done save and close the new PDF:
The script could look like

```python
rects = [evn_rect, odd_rect]
old = fitz.open("input.pdf")  # current PDF with 360 x 576 pages
new = fitz.open()  # new, target PDF
for opage in old:
    # make new page in target PDF
    npage = new.new_page(width=new_rect.width, height=new_rect.height)  
    tar_rect = rects[opage.number % 2]  # compute landing rect on new page
    npage.show_pdf_page(tar_rect, old, page.number)
new.ez_save("output.pdf")
```

**Reply by byteinsight** [reply]
That's brilliant thanks.  Works a treat and the wife is happy.    I had to swap my odd and even rects over but that might have been after I added variables rather than hardcoding but thats all.

**Reply by JorjMcKie** [reply]
Happy wife, happy life 😎
